- Add interface to get the size of the mmapped area.

baldersheim opened the pull request:
- Special handle realloc

@vekterli PR

baldersheim committed:
- Add interface to get the size of the mmapped area.
- Special handle realloc

vekterli commented on vespamalloc/src/vespamalloc/malloc/malloc.h:
```diff
@@ -235,35 +235,41 @@ void MemoryManager<MemBlockPtrT, ThreadListT>::freeSC(void *ptr, SizeClassT sc)
-        } else {
+    if (oldPtr == nullptr) return malloc(sz);
+    if ( ! _segment.containsPtr(oldPtr)) {
+        void * ptr = malloc(sz);
```
Also, presumably this transparently handles the case where _both_ the old and the new memory blocks were/are mmapped?

baldersheim replied:
Existing check for nullptr removed, as it can not happen.
Yes, both cases should be handled, and are tested.

vekterli reviewed (approved):
:+1:

baldersheim commented on vespamalloc/src/tests/test1/new_test.cpp:
```diff
@@ -148,15 +171,27 @@ detectLibrary() {
+MallocLibrary _Env = detectLibrary();
+
+size_t
+cmp(const char * v, char c, size_t count) {
```
count_mismatches it is

baldersheim commented on vespamalloc/src/vespamalloc/malloc/mmappool.cpp:
```diff
@@ -101,4 +101,11 @@ MMapPool::unmap(void * ptr) {
+MMapPool::get_size(void * ptr) const {
+    std::lock_guard guard(_mutex);
+    auto found = _mappings.find(ptr);
+    return found->second._sz;
```
Assert added.

baldersheim commented on vespamalloc/src/vespamalloc/malloc/malloc.h:
```diff
@@ -235,35 +235,41 @@ void MemoryManager<MemBlockPtrT, ThreadListT>::freeSC(void *ptr, SizeClassT sc)
+    MemBlockPtrT mem(oldPtr);
+    mem.readjustAlignment(_segment);
+    if (! mem.validAlloc()) {
+        fprintf(stderr, "Someone has tamper with my pre/post signatures of my memoryblock %p(%ld).\n", mem.ptr(), mem.size());
```
Fixed

baldersheim commented on vespamalloc/src/vespamalloc/malloc/malloc.h:
```diff
@@ -235,35 +235,41 @@ void MemoryManager<MemBlockPtrT, ThreadListT>::freeSC(void *ptr, SizeClassT sc)
-                ptr = oldPtr;
-            }
-        } else {
+    if (oldPtr == nullptr) return malloc(sz);
```
No, they have not been used since initial development and should probably just be removed.

baldersheim committed:
- Unify error handling to not check for nullptr, which can not happen.
- Assert to ensure precondition are met.
- Update length so we chcek the whole buffer.
